File: Programmers/Hash/Lv1.완주하지 못한 선수.py
from collections import Counter


def solution(participant, completion):
    participant.sort()
    completion.sort()
    answer = Counter(participant) - Counter(completion)
    return list(answer)[0]

# print(solution(["marina", "josipa", "nikola", "vinko", "filipa"], ["josipa", "filipa", "marina", "nikola"]))

'''
collection 라이브러리 사용
일반 dict는 빼기 불가,
counter를 사용하면 key, value로 되고...카운터 객체이므로 빼기 가능
마지막 리스트로 변환 후 0번째 출력--> 탈락자는 한명이다
'''

File: JS/Programmers/Hash/lv1.완주하지 못한 선수.js
function solution(participant, completion) {
    let dic = completion.reduce((obj, t) => (obj[t] = obj[t] ? obj[t] + 1 : 1, obj), {});
    console.log(dic);
    return participant.find(t => {
        if (dic[t]) {
            dic[t] = dic[t] - 1
        } else {
            return true;
        }
    })
}


const case1 = {participant: ["leo", "kiki", "eden"], completion: ["eden", "kiki"]}
const case2 = {
    participant: ["marina", "josipa", "nikola", "vinko", "filipa"],
    completion: ["josipa", "filipa", "marina", "nikola"]
}
const case3 = {participant: ["mislav", "stanko", "mislav", "ana"], completion: ["stanko", "ana", "mislav"]}

console.log(
    solution(case2.participant, case2.completion))
